File: src/parser/__tests__/fixtures/cpp/sample1.cpp
// sample1.cpp - Basic C++ features: classes, functions, structs, enums

#include <cmath>
#include <string>

// Point class representing a 2D point
class Point {
public:
    double x, y;

    Point(double x, double y) : x(x), y(y) {}

    double distanceFromOrigin() const {
        return std::sqrt(x * x + y * y);
    }

    double distanceTo(const Point& other) const {
        double dx = x - other.x;
        double dy = y - other.y;
        return std::sqrt(dx * dx + dy * dy);
    }

    static Point origin() {
        return Point(0, 0);
    }
};

// Circle class
class Circle {
public:
    static constexpr double PI = 3.14159;

    Point center;
    double radius;

    Circle(const Point& center, double radius) : center(center), radius(radius) {}

    double area() const {
        return PI * radius * radius;
    }

    double circumference() const {
        return 2 * PI * radius;
    }

    bool contains(const Point& point) const {
        return center.distanceTo(point) <= radius;
    }
};

// Calculator class
class Calculator {
private:
    double memory;

public:
    static const double MAX_VALUE;
    static const double MIN_VALUE;

    Calculator() : memory(0) {}

    double add(double a, double b) {
        memory = a + b;
        return memory;
    }

    double subtract(double a, double b) {
        memory = a - b;
        return memory;
    }

    double getMemory() const {
        return memory;
    }

    void clear() {
        memory = 0;
    }

    static double quickAdd(double a, double b) {
        return a + b;
    }
};

const double Calculator::MAX_VALUE = 1e100;
const double Calculator::MIN_VALUE = -1e100;

// ScientificCalculator inherits from Calculator
class ScientificCalculator : public Calculator {
public:
    double power(double base, double exponent) {
        return std::pow(base, exponent);
    }

    double squareRoot(double n) {
        return std::sqrt(n);
    }
};

// Struct example
struct Rectangle {
    double width;
    double height;

    double area() const {
        return width * height;
    }

    double perimeter() const {
        return 2 * (width + height);
    }
};

// Enum example
enum Status {
    Active,
    Inactive,
    Pending
};

enum class Color {
    Red,
    Green,
    Blue
};

// Constants
const int MAX_SIZE = 100;
const std::string VERSION = "1.0.0";

// Namespace example
namespace math {
    double abs(double x) {
        return x >= 0 ? x : -x;
    }

    double max(double a, double b) {
        return a > b ? a : b;
    }
}

File: src/parser/__tests__/fixtures/rust/sample1.rs
// sample1.rs - Basic Rust features: structs, enums, functions, impl blocks

/// Point in 2D space
pub struct Point {
    pub x: f64,
    pub y: f64,
}

impl Point {
    /// Create a new point
    pub fn new(x: f64, y: f64) -> Self {
        Point { x, y }
    }

    /// Calculate distance from origin
    pub fn distance_from_origin(&self) -> f64 {
        (self.x.powi(2) + self.y.powi(2)).sqrt()
    }

    /// Calculate distance to another point
    pub fn distance_to(&self, other: &Point) -> f64 {
        let dx = self.x - other.x;
        let dy = self.y - other.y;
        (dx.powi(2) + dy.powi(2)).sqrt()
    }
}

/// Circle defined by center and radius
pub struct Circle {
    center: Point,
    radius: f64,
}

impl Circle {
    /// Create a new circle
    pub fn new(center: Point, radius: f64) -> Self {
        Circle { center, radius }
    }

    /// Calculate area
    pub fn area(&self) -> f64 {
        std::f64::consts::PI * self.radius.powi(2)
    }

    /// Calculate circumference
    pub fn circumference(&self) -> f64 {
        2.0 * std::f64::consts::PI * self.radius
    }

    /// Check if point is inside circle
    pub fn contains(&self, point: &Point) -> bool {
        self.center.distance_to(point) <= self.radius
    }
}

/// Shape color
#[derive(Debug, Clone, Copy)]
pub enum Color {
    Red,
    Green,
    Blue,
    RGB(u8, u8, u8),
}

/// Calculator operations
pub enum Operation {
    Add,
    Subtract,
    Multiply,
    Divide,
}

/// Simple calculator
pub struct Calculator {
    memory: f64,
}

impl Calculator {
    /// Create a new calculator
    pub fn new() -> Self {
        Calculator { memory: 0.0 }
    }

    /// Perform an operation
    pub fn calculate(&mut self, a: f64, b: f64, op: Operation) -> f64 {
        let result = match op {
            Operation::Add => a + b,
            Operation::Subtract => a - b,
            Operation::Multiply => a * b,
            Operation::Divide => {
                if b != 0.0 {
                    a / b
                } else {
                    f64::NAN
                }
            }
        };
        self.memory = result;
        result
    }

    /// Get stored memory
    pub fn get_memory(&self) -> f64 {
        self.memory
    }

    /// Clear memory
    pub fn clear_memory(&mut self) {
        self.memory = 0.0;
    }
}

/// Maximum calculation limit
pub const MAX_CALC_VALUE: f64 = 1e100;

/// Minimum calculation limit
pub const MIN_CALC_VALUE: f64 = -1e100;

/// Default precision
pub const DEFAULT_PRECISION: usize = 6;

/// Application version
pub static VERSION: &str = "1.0.0";

/// Global counter (mutable static)
pub static mut GLOBAL_COUNTER: i32 = 0;

/// Type alias for result
pub type CalcResult = Result<f64, String>;

/// Type alias for point tuple
pub type PointTuple = (f64, f64);

/// Free function to create point from tuple
pub fn point_from_tuple(tuple: PointTuple) -> Point {
    Point::new(tuple.0, tuple.1)
}

/// Free function to calculate distance
pub fn calculate_distance(p1: &Point, p2: &Point) -> f64 {
    p1.distance_to(p2)
}

/// Helper function for safe division
fn safe_divide(a: f64, b: f64) -> CalcResult {
    if b == 0.0 {
        Err("Division by zero".to_string())
    } else {
        Ok(a / b)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_point_distance() {
        let p1 = Point::new(0.0, 0.0);
        let p2 = Point::new(3.0, 4.0);
        assert_eq!(p1.distance_to(&p2), 5.0);
    }

    #[test]
    fn test_circle_area() {
        let center = Point::new(0.0, 0.0);
        let circle = Circle::new(center, 1.0);
        assert!((circle.area() - std::f64::consts::PI).abs() < 0.001);
    }
}
